sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
+}
+----
+
+Before describing those properties those properties in more detail, a couple of
+new terms are introduced to simplify the extension specification:
+
+Set of virtual member functions:: a group of virtual member functions which are
+defined with `indirectly_callable` property with the same value of the property
+parameter `SetId`. For the simplicity, this will also be further refferred to as
+a _set_, or as a _set of virtual functions_.
+
+Kernel declares a use of a set of virtual member functions:: a kernel is
+considered to be declaring a use of a set of virtual member functions `SetIdA`
+when it is submitted with `calls_indirectly` property with `SetIdA` included
+into the property parameter `SetIds`. If `SetIdA` is not included into the
+property parameter `SetIds`, or if a kernel is submitted without the property,
+then it is *not* considered to be declaring a use of the set of virtual member
+functions.
+
+|===
+|Property|Description
+|`indirectly_callable`
+|The `indirectly_callable` property indicates that a virtual member function is
+a device function, thus making it available to be called from SYCL kernel and
+device functions. Should only be applied to virtual member functions and to do
+so, function-style `SYCL_EXT_ONEAPI_INDIRECTLY_CALLABLE_PROPERTY` macro should
+be used. The macro accepts a single optional argument, which is passed to the
+property parameter `SetId` and therefore has all the same requirements.
+
+Optional parameter `SetId` specifies a set of virtual member functions this
+function belongs to and at the same time it defines a group of kernels, which
+can call this function, it must be a C++ typename. When the parameter is
+ommitted, a virtual member function is considered to belong to the default set.
+
+Calling a virtual member function from a kernel which does not declare use of a
+set the virtual member function belongs to is an undefined behavior.
+
+|`calls_indirectly`
+|The `calls_indirectly` property indicates that a SYCL kernel function is
+performing calls through virtual member functions and declares use of one or
+more sets of virtual member functions.
+
+Optional parameter `SetIds` specifies which sets of virtual member functions are
+declared to be used by a kernel, it must be zero or more C\++ typenames. If the
+argument is omitted (zero C++ typenames specified), then a kernel is considered
+to be using the default set of virtual member functions.
+
+Calling a virtual member function, which does not belong to any of sets of
+virtual member functions declared to be used is an undefined behavior.
+|===
+
+If a virtual member function is called from device code, both definition and
+declaration of that function must be decorated with the `indirectly_callable`
+property. `SetId` property parameter must match between definition and
+declaration, or otherwise behavior is undefined.
```

Comment:
Let me do another round of prototyping here. This wording is a result of my previous experiments, where both definition and declaration had to be annotated, or otherwise some functionality was missing (either proper LLVM IR codegen, or FE diagnostics). Perhaps I missed something an prototyped it wrong and I'd done before we have agreed that we can parse properties in FE. The latter changes the landscape a bit, so with it we may be in a bit better position overall.

Reply:
Updated the wording to match #11861 in 4b622e169b8078305d130b0b02c4480b54f712b8